### PR TITLE
`rehearsal`: add supplemental images config

### PR DESCRIPTION
**Pull request by Prucek**
Add the option to supplemental images config: https://github.com/openshift/release/blob/master/core-services/image-mirroring/_config.yaml and ignore the targeting images defined there.

/cc @hongkailiu

### CODE REVIEW

File: pkg/controller/quay_io_ci_images_distributor/supplemental_images.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+	"strings"
+
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"sigs.k8s.io/yaml"
+
+	"github.com/openshift/ci-tools/pkg/api"
+	"github.com/openshift/ci-tools/pkg/util/gzip"
+)
+
+func LoadConfig(file string) (*SupplementalCIImagesConfig, error) {
```
Comment (hongkailiu) [inline pkg/controller/quay_io_ci_images_distributor/supplemental_images.go:15]:
```suggestion
func LoadConfig(file string) (*Config, error) {
```

File: cmd/pj-rehearse/main.go [hunk 2]
```diff
@@ -83,6 +84,7 @@ func gatherOptions() (options, error) {
 	fs.Var(&o.stickyLabelAuthors, "sticky-label-author", "PR Author for which the 'rehearsals-ack' label will not be removed upon a new push. Can be passed multiple times.")
 	fs.StringVar(&o.webhookSecretFile, "hmac-secret-file", "/etc/webhook/hmac", "Path to the file containing the GitHub HMAC secret.")
 	fs.StringVar(&o.registryConfig, "registry-config", "", "Path to the file of registry credentials")
+	fs.StringVar(&o.supplementalImagesConfigPath, "supplemental-images-config-path", "", "Path to the supplemental CI images config path file")
```
Comment (hongkailiu) [inline cmd/pj-rehearse/main.go:87]:
```suggestion
	fs.StringVar(&o.ciImagesMirrorConfigPath, "ci-images-mirror-config-path", "", "Path to ci-image-mirror config path file")
```

File: cmd/pj-rehearse/server.go [hunk 3]
```diff
@@ -149,6 +149,14 @@ func serverFromOptions(o options) (*server, error) {
 	}
 	rehearsalConfig.QuayIOImageHelper = quayIOImageHelper
 
+	if o.supplementalImagesConfigPath != "" {
+		if c, err := quayiociimagesdistributor.LoadConfig(o.supplementalImagesConfigPath); err != nil {
+			return nil, fmt.Errorf("failed to load config file %s: %w", o.supplementalImagesConfigPath, err)
+		} else {
+			rehearsalConfig.SupplementalCIImagesConfig = c
```
Comment (hongkailiu) [inline cmd/pj-rehearse/server.go:156]: We could just save and pass the targets of the mirroring (the keys of the mapping) onto the function.
We may call it as `ignoredTargets` (with type `sets[string]`) or something.

File: pkg/controller/quay_io_ci_images_distributor/supplemental_images.go (hunk 1 above)
Comment (Prucek) [reply pkg/controller/quay_io_ci_images_distributor/supplemental_images.go]: I can't do that, beacsue `Config`  is already defined here: https://github.com/openshift/ci-tools/blob/master/pkg/controller/quay_io_ci_images_distributor/quay_io_ci_images_distributor.go#L392

File: pkg/rehearse/rehearse.go [hunk 4]
```diff
@@ -695,12 +699,21 @@ func ensureISTSInQCI(ctx context.Context, client ctrlruntimeclient.Client, ists
...
 	colonSplit := strings.Split(ist.Name, ":")
 	if n := len(colonSplit); n != 2 {
 		return []string{}, fmt.Errorf("splitting %s by `:` didn't yield two but %d results", ist.Name, n)
 	}
 	tagRef := api.ImageStreamTagReference{Namespace: ist.Namespace, Name: colonSplit[0], Tag: colonSplit[1]}
+
+	if supplementalCIImagesConfig != nil {
+		for name, i := range supplementalCIImagesConfig.SupplementalCIImages {
+			if i.ImageStreamTagReference == tagRef || name == ist.Namespace+"/"+ist.Name {
+				log.Infof("Skipping supplemental image %s", ist.Name)
```
[the remark this answers is not in the document]
Comment (Prucek) [reply pkg/rehearse/rehearse.go]: Actually, when I think of it now, we only need `name == tagRef.Namespace+"/"+tagRef.Name+":"+tagRef.Tag`
when passing only `ignoredTargets`

File: pkg/rehearse/rehearse.go [hunk 5]
```diff
@@ -695,12 +699,19 @@ func ensureISTSInQCI(ctx context.Context, client ctrlruntimeclient.Client, ists
...
-func createPairs(ctx context.Context, ist types.NamespacedName, client ctrlruntimeclient.Client, time time.Time, log *logrus.Entry) ([]string, error) {
+func createPairs(ctx context.Context, ist types.NamespacedName, ignoredTargets sets.Set[string], client ctrlruntimeclient.Client, time time.Time, log *logrus.Entry) ([]string, error) {
 	colonSplit := strings.Split(ist.Name, ":")
 	if n := len(colonSplit); n != 2 {
 		return []string{}, fmt.Errorf("splitting %s by `:` didn't yield two but %d results", ist.Name, n)
 	}
 	tagRef := api.ImageStreamTagReference{Namespace: ist.Namespace, Name: colonSplit[0], Tag: colonSplit[1]}
+
+	for name := range ignoredTargets {
+		if name == tagRef.Namespace+"/"+tagRef.Name+":"+tagRef.Tag {
```
Comment (hongkailiu) [inline pkg/rehearse/rehearse.go:710]:
```suggestion
		if name == tagRef.ISTagName() {
```

File: pkg/rehearse/rehearse.go [hunk 6]
```diff
@@ -695,12 +699,19 @@ func ensureISTSInQCI(ctx context.Context, client ctrlruntimeclient.Client, ists
...
+			log.Infof("Skipping supplemental image %s", ist.Name)
```
Comment (hongkailiu) [inline pkg/rehearse/rehearse.go:711]:
```suggestion
			log.WithField("tagRef", tagRef.ISTagName()).Info("Skipping a ignored image")
```

**Review by hongkailiu — COMMENTED** [review]
Only a couple of minors.
Otherwise, LGTM.

File: pkg/rehearse/rehearse.go [hunk 7]
```diff
@@ -695,12 +699,19 @@ func ensureISTSInQCI(ctx context.Context, client ctrlruntimeclient.Client, ists
...
+func createPairs(ctx context.Context, ist types.NamespacedName, ignoredTargets sets.Set[string], client ctrlruntimeclient.Client, time time.Time, log *logrus.Entry) ([]string, error) {
 	colonSplit := strings.Split(ist.Name, ":")
 	if n := len(colonSplit); n != 2 {
 		return []string{}, fmt.Errorf("splitting %s by `:` didn't yield two but %d results", ist.Name, n)
 	}
 	tagRef := api.ImageStreamTagReference{Namespace: ist.Namespace, Name: colonSplit[0], Tag: colonSplit[1]}
+
+	for name := range ignoredTargets {
+		if name == tagRef.ISTagName() {
+			log.WithField("tagRef", tagRef.ISTagName()).Info("Skipping a ignored image")
```
Comment (hongkailiu) [inline pkg/rehearse/rehearse.go:711]: My bad

```suggestion
			log.WithField("tagRef", tagRef.ISTagName()).Info("Skipping an ignored image")
```

**Pull request merged**